crates/media-parser/src/format/mp3/duration.rs

```diff
@@ -0,0 +1,482 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Duration {
+   /// Duration in milliseconds.
+   pub millis: u64,
+   /// Method used for calculation.
+   pub method: DurationMethod,
+}
+
+impl Duration {
+   /// Creates a new duration value.
+   pub const fn new(millis: u64, method: DurationMethod) -> Self {
+      Self { millis, method }
+   }
+
+   /// Returns duration in seconds as floating point.
+   pub fn seconds(&self) -> f64 {
+      self.millis as f64 / 1000.0
+   }
+
+   /// Returns zero duration with estimated method.
+   pub const fn zero() -> Self {
+      Self {
+         millis: 0,
+         method: DurationMethod::Estimated,
+      }
+   }
+}
+
+/// Method used for duration calculation.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DurationMethod {
+   /// Calculated from Xing/VBRI header (exact for VBR files).
+   VbrHeader,
+   /// Calculated from file size and bitrate (exact for CBR files).
+   Cbr,
+   /// Estimated (fallback when other methods fail).
+   Estimated,
+}
+
+/// VBR header information (Xing or VBRI).
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct VbrInfo {
+   /// Total number of frames in file.
+   pub total_frames: u32,
+   /// Total bytes of audio data (optional).
+   pub total_bytes: Option<u32>,
+   /// VBR header type.
+   pub header_type: VbrHeaderType,
+}
+
+/// Type of VBR header found.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum VbrHeaderType {
+   /// Xing header (common in LAME encoded files).
+   Xing,
+   /// Info header (CBR files encoded with LAME).
+   Info,
+   /// VBRI header (Fraunhofer encoder).
+   Vbri,
+}
+
+/// Trait for duration calculation options.
+///
+/// Enables polymorphism and dependency injection for different
+/// calculation methods.
+pub trait DurationStrategy: Send + Sync {
+   /// Calculates duration from available information.
+   ///
+   /// # Parameters
+   ///
+   /// - `first_frame`: Parsed header of first audio frame
+   /// - `audio_start`: Byte offset where audio data starts
+   /// - `file_size`: Total file size in bytes
+   /// - `vbr_info`: Optional VBR header information
+   fn calculate(
+      &self,
+      first_frame: &FrameHeader,
+      audio_start: u64,
+      file_size: u64,
+      vbr_info: Option<&VbrInfo>,
+   ) -> Duration;
+}
+
+/// CBR (Constant Bit Rate) duration strategy.
+///
+/// Calculates duration from file size and bitrate:
+/// `duration = audio_bytes * 8 / bitrate`
+pub struct CbrStrategy;
+
+impl DurationStrategy for CbrStrategy {
+   fn calculate(
+      &self,
+      first_frame: &FrameHeader,
+      audio_start: u64,
+      file_size: u64,
+      _vbr_info: Option<&VbrInfo>,
+   ) -> Duration {
+      if first_frame.bitrate_kbps == 0 || file_size <= audio_start {
+         return Duration::zero();
+      }
+
+      let audio_bytes = file_size - audio_start;
+      let bitrate_bps = first_frame.bitrate_kbps as u64 * 1000;
+      let millis = (audio_bytes * 8 * 1000) / bitrate_bps;
```

Comment:
Reorder the math to divide first, avoiding a _potentially_ large intermediate value which could cause the `u64` to overflow.